packages/core/src/fusion/Map.ts

```diff
@@ -42,13 +46,15 @@ export default class Map {
   }
 }
 
-class MapSink extends Pipe {
-  constructor (f, sink) {
+class MapSink<A, B> extends Pipe<A | B> implements Sink<A | B> {
```

Comment:
I think this should just be `extends Pipe<A>`:

1. It shouldn't `extends Pipe<A | B>`.  The intent is that MapSink accepts `A`s, so the interface should be, e.g. `event (t: Time, a: A): void`.  It uses the provided function to turn `A`s into `B`s and passes them to the provided `Sink<B>`.
2. I think we should just `extends Pipe<A>`, and not also `implements Sink<A>`.  Any subclass of Pipe must still implement Sink, and the type system will catch any problems.  Is there a concrete advantage to adding the redundant `implements Sink<A>`?

Reply:
Union type is need to satisfy both `super` call in constructor and `event` handler. Constructor accepts `sink: Sink<B>` and sends in to parent `Pipe<B>` - this explains `extends Pipe<B>`. On the other hand `MapSink` is itself a `Sink<A>` and therefore it should handle `event(t: Time, x: A): void` - but it conflicts with inherited `event(t: Time, x: B): void` from `Pipe<B>`. That's why we should either remove the parent `Pipe` or define the type as a union `A | B`. This applies to all extensions of `Pipe`.

Reply:
Ah, yes, thanks.  That's a real bummer, and probably a design flaw in my using Pipe just as a way to provide default implementations of some methods and reduce verbosity.  It's still the case, though, that `implements Sink<A | B>` doesn't accurately reflect the intent of MapSink.  It should only accept `A`s.  Hmmm ...

It looks like every place that extends Pipe currently provides an implementation of `event` (but not always `end` and `error`).  If that's the case, then perhaps a solution is for Pipe to be abstract (which seems to match the intent as well) and not provide an implementation of `event`, like:

```ts
export default abstract class Pipe<A, B> implements Sink<A> {
  protected readonly sink: Sink<B>

  constructor(sink: Sink<B>) {
    this.sink = sink
  }

  abstract event(t: Time, x: A): void;

  end(t: Time): void {
    return this.sink.end(t)
  }

  error(t: Time, e: Error): void {
    return this.sink.error(t, e)
  }
}
```

That is, the mechanism by which `Pipe<A, B>` turns `A`s into `B`s is unspecified.  This still allows `extends Pipe<A, A>`, where applicable, which makes the nice guarantee that it _won't_ change the event value type.  I think this is a non-breaking change, since it looks like Pipe has never been exported from `@most/core`.

The MapSink case would then become:

```ts
class MapSink<A, B> extends Pipe<A, B> {
  constructor(private readonly f: (a: A) => B, sink: Sink<B>) {
    super(sink)
    this.f = f
  }

  event(t: Time, x: A): void {
    const f = this.f
    this.sink.event(t, f(x))
  }
}
```

That seems to eliminate the need for the union type (by separating the input and output types), and retain the ability to provide default implementations of `end` and `error` for convenience.

Thoughts?  Other options?

If that doesn't work, then I may just be in favor of ditching Pipe and only using `implements Sink<A>`.

Reply:
Nice, it works, thanks

Reply:
Yay! :+1: